Comments: Add edit form

Copons opened the pull request:
Close #13905

Introduce the Comment Management Edit Form and enable it to development, wpcalypso, and horizon environments.

![sep-04-2017 18-50-45](https://user-images.githubusercontent.com/2070010/30035536-1bf47c96-91a3-11e7-96cf-ac220673241c.gif)

| Not Edit Mode | Edit Mode | Edit Mode (Registered User) |
| --- | --- | --- |
| <img width="682" alt="screen shot 2017-08-17 at 11 02 51" src="https://user-images.githubusercontent.com/2070010/29407257-bc3ec682-833b-11e7-9bb2-2a98250abbdc.png"> | <img width="597" alt="screen shot 2017-09-05 at 12 31 31" src="https://user-images.githubusercontent.com/2070010/30059240-2f1d93d2-9236-11e7-9f91-0e28aca6d915.png"> | <img width="597" alt="screen shot 2017-09-05 at 12 32 30" src="https://user-images.githubusercontent.com/2070010/30059281-549e94bc-9236-11e7-83c0-589a5803fa1f.png"> |

Clicking on the **Edit** button in the action bar activates the Edit Mode.
In Edit Mode the actions are replaced by a simple "Edit Comment" text, and the Collapse button is replaced by an X button.
The only ways to exit Edit Mode are:
- Clicking the X button that replaces the Collapse button.
- Clicking the Cancel button.
- Clicking the Save button (which also updates the comment).

The author name and URL fields are only enabled for unregistered users.

On WP.com and JP > 5.3 sites, the editable content is the comment's `raw_content`.
Otherwise it's the formatted `content`.

# Action Items

- [x] Add the author name and URL to the editable fields in the endpoint.

- [x] Replace the `editComment` action with a data layer one, to take better advantage of the whole success/error flow, with optimistic updates and notices. (https://github.com/Automattic/wp-calypso/pull/17252)

- [x] Find out how to juggle `context: 'display'` and `context: 'edit'` in the request, to switch between raw and rendered content, when the Edit Mode is respectively enabled and disabled. (see: https://github.com/Automattic/wp-calypso/pull/17251#issuecomment-322829231 and following comments)

gwwar commented:
We should also add a feature flag for editing. This will lets us land wip PRs

gwwar commented:
Awesome work so far @copons!

I noticed a corner case, that we could file in a separate bug.

But when we edit with and add some input that will be sanitized, our optimistic update won't necessarily match what the server returns:

<img width="942" alt="screen shot 2017-08-21 at 3 18 30 pm" src="https://user-images.githubusercontent.com/1270189/29540683-95bf10f6-8684-11e7-97dd-d25c56a13681.png">

After Edit:
<img width="941" alt="screen shot 2017-08-21 at 3 24 15 pm" src="https://user-images.githubusercontent.com/1270189/29540735-d82521f6-8684-11e7-973d-9dae56c3cde6.png">

Page Refresh:
<img width="942" alt="screen shot 2017-08-21 at 3 18 30 pm" src="https://user-images.githubusercontent.com/1270189/29540744-df73ba08-8684-11e7-803e-96833d7562b5.png">




Also, adding a needs design label, for help.

gwwar commented:
Looking at this again we can 🚢 this after:
- [ ] adding a feature flag
- [ ] verifying that we have no regressions

I think this is a good chunk of work! We can then continue on this in more incremental PRs.

Copons commented:
> When we edit with and add some input that will be sanitized, our optimistic update won't necessarily match what the server returns:

@gwwar IIRC this wold be fixed by https://github.com/Automattic/wp-calypso/pull/17252

gwwar commented:
> Might be worth opening a low priority issue

That's what I'm planning to do. I don't think it should be too bad to correct. Once the success response comes back we can dispatch and update the local comment if it doesn't match current content.

folletto commented:
Question: would be possible to move the "Cancel" and "Save" buttons to replace entirely the controls in the header? So the whole list (x, like, approve, edit, spam, trash) disappears, and we just have the buttons at the top?

gwwar commented:
> So the whole list (x, like, approve, edit, spam, trash) disappears, and we just have the buttons at the top?

Interesting thought @folletto. We can certainly try that out and see how it feels. 

I do have a few minor concerns with the change. Adding another state to the top action bar makes the component more complex. There's also longer mouse travel after modifying content. This may also cause trouble in smaller views.

folletto commented:
>  This may also cause trouble in smaller views.

I felt it would be better in smaller views, as the top bar could stick when opened. What's the concern?

> I do have a few minor concerns with the change. Adding another state to the top action bar makes the component more complex. 

If this is a concern no worries, we can review later.

megsfulton commented:
Took a closer look at the Edit behavior and agree that it's a little off to have the other icons stick around while in edit mode. 

I mocked up two options for us to consider that take into account the UI changes we made earlier of moving the icons to the left w/ the chevron on the right. 

Option A
- Creates a clear state change between normal mode and edit mode
- Replaces the chevron with an X as an additional way to cancel out of edit mode 

Option B
- Based on @folletto 's suggestion and moves the Cancel / Save buttons to the top
- Uses a similar state change from Option A
- One thing to consider is on mobile web when the comment field is focused with the keyboard up is that there's the possibility of the Edit/Cancel/Save bar getting pushed out of view (see example here: https://cloudup.com/cLBOWSbohJs)  

Option C - not pictured 
- Both Options A & B introduce a new state change which adds extra complexity soooo if we're particularly concerned about users get stuck, another option would be to have clicking or tapping in any area of the top action bar with an icon cancels out of edit mode 

<img width="886" alt="edit comment" src="https://user-images.githubusercontent.com/161697/29798366-75db79f4-8c11-11e7-857a-1bc1379fb61d.png">

gwwar commented:
Thanks @megsfulton! I'm personally biased toward Option A, since I feel like our eyes move from top to bottom in the edit flow, and it's pretty natural to have the save button on the bottom.

> here's the possibility of the Edit/Cancel/Save bar getting pushed out of view

If we still want buttons on the top, might I suggest putting both save/cancel buttons on the top + bottom?

megsfulton commented:
>If we still want buttons on the top, might I suggest putting both save/cancel buttons on the top + bottom?

Hmm...that's a lot of buttons for a small space 😅 

I'm a fan of option A too. I think it makes it clear that you're in an editing mode and keeps the save button closer to the fields that are being edited.

gwwar commented:
To summarize, what we chatted about on Slack, for each of these Jetpack endpoint updates

* wpcom  + jetpack with >= 5.3 can use the new approach
* everything else uses the fallback method (or current approach)

gwwar commented:
@folletto any blockers for us running with Option A here?

folletto commented:
I've still a preference for B, but it's minor now compared to A.

Go ahead with A. :)

Copons commented on client/state/comments/reducer.js:
```diff
@@ -50,15 +50,35 @@ export const deconstructStateKey = key => {
+				},
+				content: newProperties.commentContent,
+			}
+		: { ...comment, ...newProperties };
```
This weird indentation was caused by the automatic Prettier on commit.
It throws an eslint warning, but I'm not sure how I'm supposed to fix it.
I don't feel like removing the `/** @format */` marker, so should I change this into an if/else?

Copons commented on client/state/comments/reducer.js:
```diff
@@ -50,15 +50,35 @@ export const deconstructStateKey = key => {
+					url: newProperties.authorUrl,
+				},
+				content: newProperties.commentContent,
+			}
```
This weird indentation was caused by the automatic Prettier on commit.
It throws an eslint warning, but I'm not sure how I'm supposed to fix it.
I don't feel like removing the `/** @format */` marker, so should I change this into an if/else?

Copons commented:
Update time!
I've wired everything up and refreshed the design according to the suggestions. Here's the result:

![sep-04-2017 18-50-45](https://user-images.githubusercontent.com/2070010/30035536-1bf47c96-91a3-11e7-96cf-ac220673241c.gif)

Stuff to consider / worth noticing or discussing:

- The "Edit Comment" title might look like an action button? Consider how the actions look like on a large screen, when they are displayed with the labels.

- I've chosen to not hide the "Edit Comment" label on small screen (as it happens for the actions), as I believe there's enough space for it.

- The raw content is only used for WP.com and JP>5.3 sites. Everything else will modify the formatted content instead. Is this what we want @gwwar?

- The comment is optimistically updated. You can notice it in the gif: upon save, the content contains an "un-emojified" `:)`. When the POST request succeed, the content is replaced with the actual formatted content returned by the API, and the `:)` is replaced by `🙂`.
(Please don't mind the fact that the emoji looks wrong. It's another issue - already filed.)

- There is a field names mismatch between GET response/Redux state and POST request.
I've opted to do two conversions, and failing to do so would lead to information loss during the optimistic update (because the API expects the author name as `author`, which would overwrite the entire `author` object in the reducer).
  1. In the reducer: we check if the action data contains the specific field names sent by Comment Management's Edit form, and in this case we convert the fields and merge them into the comment object.
  2. In the data layer function: again we check if we're coming from Comment Management, and in this case we properly name the fields as expected by the API.

--- 

- Not shown in the gif is the behaviour of modifying the author name and URL:
  - If the author is **not** a WP.com registered user, we can modify it no problem.
  - Otherwise, we just can't. The optimistic update would show the updated name/URL, but the server response would "revert" it soon after.

I guess we have two options here:
- Remove the author fields.
- Disable the author fields.

I'm personally oriented to the second option, maybe with a little message saying something like "This is a registered user, you cannot modify their name and URL".

It would feel a little less jarring than having some comments with 3 fields and some with just 1.

folletto commented:
> The "Edit Comment" title might look like an action button? Consider how the actions look like on a large screen, when they are displayed with the labels.

Yes, good point.
We could either remove the icon, remove the label, or keep both and user a light gray like it was disabled (in the end, it's just decoration text).

> Disable the author fields.

Yes. Just disable.

rodrigoi commented on client/state/comments/reducer.js:
```diff
@@ -50,15 +50,35 @@ export const deconstructStateKey = key => {
+	has( newProperties, 'authorDisplayName' ) &&
+	has( newProperties, 'authorUrl' );
+
 const updateComment = ( commentId, newProperties ) => comment => {
```
I think this function has grown to the point that it justifies having a test suite, specially with the split behaviour of edit vs like/status change.

Copons replied:
Good point!
Added a new test case targeting the three Comment Management fields (`commentContent`, `authorDisplayName`, `authorUrl`) in f175eef08473d3a49ee197441365e0c60d9fc53d

kwight commented:
While there are still a few issues and decisions here (editing registered users, different content formats for Jetpack versions, supporting older versions at all), this works great, is tucked behind a flag, and is very very big – I'm with @gwwar , let's land and continue in other PRs 👍

kwight commented:
Just noting that I've come full-circle in thinking of editing names and URLs of registered users: let's not do it. While `wp-admin` allows this, it's only adding more complexity for the sake of an odd (and, I'm imagining, rarely used) feature.

Copons commented:
@kwight All right then!
I'm gonna ship this and keep iterating in other PRs.

Just as a heads up before, I've pushed 2 changes:

- Changed the naming of the various update and edit functions.
Oddly, at first I called `edit` the function responsible to toggle the edit mode. For this reason I started calling the actual edit function `updateComment`.
In the end, it doesn't really make sense, and it's inconsistent with the Redux action and the data layer.
Therefore, I've renamed `edit` into `toggleEditMode`, and `updateComment` into `editComment`.

- I've fixed a bug where if the Comment Management Edit form sends one or more empty fields, the check in the data layer fails because it only looks for truthy values:
`if ( comment.authorDisplayName && comment.authorUrl && comment.commentContent )`
Since those three fields can only be sent from the Comment Management Edit form, checking for all three is unnecessary.
So, I've simply replaced the `&&`s with `||`s.
